fem/field_interpolant.cpp

```diff
@@ -0,0 +1,441 @@
+      inv.Factor(mi);
+      const FiniteElement &fe = *fes.GetFE(e);
+      ElementTransformation &eltr = *fes.GetElementTransformation(e);
+      qi.AssembleRHSElementVect(fe, eltr, rhs);
+      for (int ind = 0; ind < vdim; ind++)
+      {
+         qfv_sub.MakeRef(qfv, ndofs * ind);
+         rhs_sub.MakeRef(rhs, ndofs * ind);
+         inv.Mult(rhs_sub, qfv_sub);
+      }
+      fes.GetElementVDofs(e, dofs);
+      gf.SetSubVector(dofs, qfv);
+   }
+}
+
+void FieldInterpolant::ProjectQuadratureDiscCoefficient(GridFunction &gf,
+                                                        QuadratureFunctionCoefficient &qfc,
+                                                        FiniteElementSpace &fes)
+{
+   int ndofs;
+   DenseMatrix mi;
+   DenseMatrixInverse inv(&mi);
+   const IntegrationRule* ir;
+   {
+      // This is the best way I can think of to make sure the IntegrationRule in the FiniteElementSpace
+      // and the QuadratureSpace correspond to the same
+      const FiniteElement &el = *fes.GetFE(0);
+      ir = &(IntRules.Get(el.GetGeomType(), 2 * el.GetOrder() + 1));
+      const QuadratureFunction* qf = qfc.GetQuadFunction();
+      const IntegrationRule* ir_qf = &qf->GetSpace()->GetElementIntRule(0);
+      MFEM_VERIFY((ir->GetOrder() == ir_qf->GetOrder()) &&
+                  (ir->GetNPoints() == ir_qf->GetNPoints()),
+                  "IntegrationRule in FiniteElementSpace and in QuadratureFunction appear to be different");
+      // This should be the number of nodes available
+      ndofs = el.GetDof();
+   }
+
+   Vector rhs(ndofs);
+   Vector qfv(ndofs);
+   Array<int> dofs;
+
+   QuadratureIntegrator qi(qfc);
+   qi.SetIntRule(ir);
+   Array<BilinearFormIntegrator*> dbfi = *L2->GetDBFI();
+
+   if (!setup_disc)
+   {
+      m_all_data.SetSize(ndofs * ndofs * NE);
+      double* data = m_all_data.HostReadWrite();
+      for (int e = 0; e < NE; e++)
+      {
+         const FiniteElement &fe = *fes.GetFE(e);
+         ElementTransformation &eltr = *fes.GetElementTransformation(e);
+         mi.UseExternalData((data + (ndofs * ndofs * e)), ndofs, ndofs);
+         dbfi[0]->AssembleElementMatrix(fe, eltr, mi);
+      }
+      setup_disc = true;
+   }
+
+
+
+   double* data = m_all_data.HostReadWrite();
+   for (int e = 0; e < NE; e++)
+   {
+      mi.UseExternalData((data + (ndofs * ndofs * e)), ndofs, ndofs);
+      inv.Factor(mi);
+      const FiniteElement &fe = *fes.GetFE(e);
+      ElementTransformation &eltr = *fes.GetElementTransformation(e);
+      qi.AssembleRHSElementVect(fe, eltr, rhs);
+      inv.Mult(rhs, qfv);
+      fes.GetElementDofs(e, dofs);
+      gf.SetSubVector(dofs, qfv);
+   }
+}
+
+void FieldInterpolant::ProjectQuadratureCoefficient(GridFunction &gf,
+                                                    VectorQuadratureFunctionCoefficient &vqfc,
+                                                    FiniteElementSpace &fes)
+{
+   const IntegrationRule* ir;
```

Comment:
Why do you create a new `IntegrationRule` and check if it's the correct one? Can't you just use the correct one from the `QuadratureFunction`?